<img src="https://raw.githubusercontent.com/carubbi/MQ/b58af070c45fb17b77640c06c0016859ad366b4d/notebooks/assets/imgs/UNIFOR_logo.png" width="400">
<br>
<b>
<font size="6" face="arial" color="blue">
    Graduação em Ciência da Computação
</font>
</b>
<br>
<b>
<font size="4" face="arial">
    Disciplina: Métodos Quantitativos em Computação
</font>
</b>

**Orientador: Prof. Me. Ricardo Carubbi** <br>
*Docente da Graduação e Pós-Graduação em Ciência de Dados e Inteligência Artificial*<br>
*Laboratório de Ciência de Dados e Inteligência Artificial*<br>
*Universidade de Fortaleza*<br>

# **Aula 5 — Tipos, qualidade e pré-processamento básico**

## 1. Objetivos de aprendizagem

Ao final da aula, o aluno será capaz de:

- distinguir unidade de análise, observação, variável, identificador e metadado;
- classificar variáveis por sua natureza estatística e seu papel analítico, considerando significado, processo de obtenção e granularidade do registro;
- diferenciar tipos estatísticos de tipos computacionais e avaliar a coerência de conversões;
- diagnosticar completude, validade, consistência e unicidade com denominadores e chaves adequados;
- justificar transformações básicas que preservem a base bruta e a rastreabilidade das decisões.

## 2. Agenda

1. Unidade de análise e papel das variáveis — 15 min
2. Natureza estatística das variáveis — 25 min
3. Tipo estatístico versus tipo computacional — 15 min
4. Qualidade dos dados orientada pela pergunta — 20 min
5. Pré-processamento, chaves e rastreabilidade — 20 min
6. Estudo e exercícios — 5 min

## 3. Unidade de análise e papel das variáveis

### 3.1 Unidade de análise, observação e variável

Antes de classificar uma coluna, é necessário definir o que cada linha representa. A **unidade de análise** é a entidade sobre a qual se deseja produzir uma descrição ou conclusão. Uma **observação** é o registro dessa unidade na base, enquanto uma **variável** é uma característica que pode assumir valores diferentes entre as observações. Em uma matriz de dados organizada, as linhas representam casos e as colunas representam variáveis (BARBETTA; REIS; BORNIA, 2010, pp. 52–53).

No conjunto Palmer Penguins, o significado da linha depende do protocolo de coleta. Não basta dizer que uma linha é “um pinguim”: é preciso verificar se o mesmo indivíduo pode aparecer em campanhas diferentes e quais campos identificam cada registro. Essa definição orientará posteriormente a escolha das chaves e a análise de duplicidades.

Uma coluna pode funcionar como identificador, distinguindo entidades ou registros, ou como metadado, descrevendo o contexto da coleta. Esses papéis dependem da pergunta: uma data de coleta, por exemplo, também pode ser analisada para investigar diferenças entre períodos.

### 3.2 Processo de obtenção, unidade, precisão e granularidade

A classificação estatística depende do significado e do processo de obtenção. É necessário perguntar se o valor foi contado, medido, codificado ou agrupado; qual é sua unidade; qual precisão foi registrada; e qual é a granularidade do registro. Uma idade registrada apenas em anos aparece como inteiro, embora o tempo seja conceitualmente contínuo. Da mesma forma, uma massa medida em gramas continua sendo quantitativa contínua mesmo quando o instrumento registra valores arredondados (PINHEIRO et al., 2009, pp. 24–26).

**Aplicação guiada:** para cada coluna selecionada do Palmer Penguins, responda: o que representa, como foi obtida, qual é sua unidade ou domínio e qual papel exerce na análise?

In [2]:
# Carrega a base bruta sem aplicar transformações
import pandas as pd

arquivo_dados = "https://raw.githubusercontent.com/carubbi/MQ/main/data/raw/penguins_raw.csv"
penguins = pd.read_csv(arquivo_dados)
penguins.head()

,studyName,Sample Number,Species,Region,Island,Stage,Individual ID,Clutch Completion,Date Egg,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Sex,Delta 15 N (o/oo),Delta 13 C (o/oo),Comments
0,PAL0708,1,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N1A1,Yes,2007-11-11,39.1,18.7,181.0,3750.0,MALE,NaN,NaN,Not enough blood for isotopes.
1,PAL0708,2,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N1A2,Yes,2007-11-11,39.5,17.4,186.0,3800.0,FEMALE,8.94956,-24.69454,NaN
2,PAL0708,3,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N2A1,Yes,2007-11-16,40.3,18.0,195.0,3250.0,FEMALE,8.36821,-25.33302,NaN
3,PAL0708,4,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N2A2,Yes,2007-11-16,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Adult not sampled.
4,PAL0708,5,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N3A1,Yes,2007-11-16,36.7,19.3,193.0,3450.0,FEMALE,8.76651,-25.32426,NaN


**Tabela 1 - Primeiras observações da base bruta.** Cada linha registra uma observação da coleta, enquanto as colunas combinam características do pinguim, identificadores e metadados. A inspeção inicial ajuda a formular perguntas, mas não substitui o diagnóstico do conjunto completo. Fonte: Palmer Penguins.

In [3]:
# Resume armazenamento, preenchimento e exemplos de cada coluna
inventario = pd.DataFrame({
    "variavel": penguins.columns,
    "dtype_inicial": penguins.dtypes.astype(str).to_numpy(),
    "nao_ausentes": penguins.notna().sum().to_numpy(),
    "ausentes": penguins.isna().sum().to_numpy(),
    "valores_distintos": penguins.nunique(dropna=True).to_numpy(),
})
inventario

,variavel,dtype_inicial,nao_ausentes,ausentes,valores_distintos
0,studyName,str,344,0,3
1,Sample Number,int64,344,0,152
2,Species,str,344,0,3
3,Region,str,344,0,1
4,Island,str,344,0,3
5,Stage,str,344,0,1
6,Individual ID,str,344,0,190
7,Clutch Completion,str,344,0,2
8,Date Egg,str,344,0,50
9,Culmen Length (mm),float64,342,2,164


**Tabela 2 - Inventário estrutural do Palmer Penguins.** A base contém 344 observações e 17 colunas com diferentes quantidades de valores ausentes e distintos. Os tipos exibidos descrevem o armazenamento inicial; a natureza estatística ainda depende do significado e do processo de obtenção de cada variável.

## 4. Natureza estatística das variáveis

### 4.1 Qualitativas nominais e binárias

Variáveis **qualitativas nominais** registram categorias sem ordem natural. Espécie e ilha são exemplos: seus valores podem ser comparados por igualdade e resumidos por frequências, mas não por diferenças numéricas. Códigos numéricos usados para representar categorias continuam nominais quando os números funcionam apenas como rótulos.

Uma variável **binária** possui duas categorias e constitui um caso particular das variáveis qualitativas. O destaque é útil porque respostas como presença/ausência ou sim/não aparecem com frequência, mas os códigos 0 e 1 não transformam automaticamente essas categorias em medidas quantitativas (MORETTIN; BUSSAB, 2010, pp. 26–27).

### 4.2 Qualitativas ordinais

Variáveis **qualitativas ordinais** possuem categorias com ordem natural, como baixo, médio e alto. A ordem permite comparações de posição, porém não garante que a distância entre categorias consecutivas seja mensurável ou constante. A ordenação deve decorrer do significado da variável e, quando usada no software, precisa ser declarada explicitamente.

### 4.3 Quantitativas discretas e contínuas

Variáveis **quantitativas discretas** são produzidas tipicamente por contagem e assumem valores separados, como o número de ocorrências. Variáveis **quantitativas contínuas** resultam de mensuração e podem, conceitualmente, assumir qualquer valor em um intervalo, como massa ou comprimento. A precisão finita do instrumento não altera essa natureza (MORETTIN; BUSSAB, 2010, pp. 26–27).

### 4.4 Operações e representações coerentes

O tipo estatístico delimita operações interpretáveis. Frequências e proporções são adequadas para categorias; ordenações são pertinentes para categorias ordinais; diferenças, médias e medidas de dispersão exigem valores quantitativos com significado numérico. Na Aula 6, essa classificação orientará a escolha entre tabelas de frequências e representações gráficas. O diagrama de dispersão será mantido como representação de pares de variáveis quantitativas, sem antecipar sua análise bivariada.

| Variável | Natureza estatística | Operações coerentes |
| --- | --- | --- |
| `Species` | qualitativa nominal | frequências, proporções e comparação de categorias |
| `Sex` | qualitativa binária no registro disponível | frequências e proporções, considerando ausências |
| `Sample Number` | identificador sequencial | unicidade e rastreamento; não média |
| `Body Mass (g)` | quantitativa contínua | diferenças, medidas-resumo e histogramas |
| `Date Egg` | temporal | ordenação, intervalos e agrupamentos temporais |

In [4]:
# Contrasta aparência, armazenamento e papel analítico
variaveis_selecionadas = [
    "Species",
    "Individual ID",
    "Sample Number",
    "Body Mass (g)",
    "Date Egg",
]
perfil_variaveis = pd.DataFrame({
    "variavel": variaveis_selecionadas,
    "dtype": [str(penguins[coluna].dtype) for coluna in variaveis_selecionadas],
    "exemplo": [penguins[coluna].dropna().iloc[0] for coluna in variaveis_selecionadas],
    "distintos": [penguins[coluna].nunique() for coluna in variaveis_selecionadas],
})
perfil_variaveis

,variavel,dtype,exemplo,distintos
0,Species,str,Adelie Penguin (Pygoscelis adeliae),3
1,Individual ID,str,N1A1,190
2,Sample Number,int64,1,152
3,Body Mass (g),float64,3750.0,94
4,Date Egg,str,2007-11-11,50


**Tabela 3 - Aparência e armazenamento de cinco variáveis.** `Sample Number` aparece como inteiro, mas funciona como identificador sequencial; `Date Egg` chega como texto, embora represente uma data. Esses contrastes mostram por que o `dtype` não determina sozinho a classificação estatística.

In [5]:
# Aplica operações compatíveis com o significado das variáveis
operacoes_coerentes = pd.DataFrame([
    {
        "variavel": "Species",
        "operacao": "frequência da categoria mais comum",
        "resultado": int(penguins["Species"].value_counts().iloc[0]),
    },
    {
        "variavel": "Individual ID",
        "operacao": "quantidade de identificadores distintos",
        "resultado": int(penguins["Individual ID"].nunique()),
    },
    {
        "variavel": "Body Mass (g)",
        "operacao": "massa média observada, em gramas",
        "resultado": round(penguins["Body Mass (g)"].mean(), 2),
    },
])
operacoes_coerentes

,variavel,operacao,resultado
0,Species,frequência da categoria mais comum,152.00
1,Individual ID,quantidade de identificadores distintos,190.00
2,Body Mass (g),"massa média observada, em gramas",4201.75


**Tabela 4 - Operações coerentes com o significado das variáveis.** Frequências descrevem espécies, a contagem de valores distintos auxilia a avaliação de identificadores e a média possui interpretação física para massa corporal. A mesma média não seria informativa para códigos ou identificadores.

## 5. Tipo estatístico versus tipo computacional

### 5.1 Significado estatístico e `dtype`

O **tipo estatístico** descreve o significado da variável e as operações que fazem sentido. O **tipo computacional** informa como o software armazena e processa os valores. Eles se relacionam, mas não existe correspondência biunívoca: uma coluna inteira pode conter códigos nominais; uma contagem com ausências pode ser armazenada como ponto flutuante; e uma data pode chegar como texto. A tipagem auxilia o processamento, mas não substitui a interpretação conceitual (BRUCE; BRUCE, 2019, pp. 23–26).

### 5.2 Categorias, números, datas e valores ausentes

Em pandas, `object` frequentemente armazena texto, mas também pode reunir conteúdos heterogêneos. `category` explicita um domínio categórico; categorias ordinais precisam ter sua ordem definida. Tipos inteiros e de ponto flutuante representam números, enquanto `datetime64` representa instantes ou datas. Valores ausentes podem modificar o `dtype` observado sem modificar a natureza estatística da variável.

Assim, a classificação deve seguir esta ordem: compreender o significado, examinar os valores e o processo de obtenção, definir a natureza estatística e somente então avaliar se o tipo computacional é adequado.

### 5.3 Conversões com controles antes e depois

Uma conversão não é correta apenas porque o comando foi executado sem erro. Antes de converter, registre número de linhas, ausências, domínios e formatos problemáticos. Depois, repita os controles e compare os resultados. A conversão deve preservar o significado e tornar explícitas eventuais perdas.

Para datas, valores que não puderem ser interpretados precisam ser identificados; para categorias, níveis e frequências devem ser preservados; para números, valores inválidos introduzidos pela conversão devem ser contabilizados. Esse procedimento separa uma transformação tecnicamente possível de uma decisão analiticamente justificável.

In [6]:
# Converte a data em uma cópia e compara as perdas
dados = penguins.copy(deep=True)
ausentes_data_antes = int(dados["Date Egg"].isna().sum())
dados["Date Egg"] = pd.to_datetime(dados["Date Egg"], errors="coerce")
controle_data = pd.DataFrame({
    "controle": ["linhas", "ausentes antes", "ausentes depois", "data mínima", "data máxima"],
    "resultado": [
        len(dados),
        ausentes_data_antes,
        int(dados["Date Egg"].isna().sum()),
        dados["Date Egg"].min().date(),
        dados["Date Egg"].max().date(),
    ],
})
controle_data

,controle,resultado
0,linhas,344
1,ausentes antes,0
2,ausentes depois,0
3,data mínima,2007-11-09
4,data máxima,2009-12-01


**Tabela 5 - Controle da conversão de `Date Egg`.** A conversão preserva as 344 observações, não introduz datas ausentes e recupera o intervalo de 2007 a 2009. O controle evidencia a mudança de representação computacional sem perda observada.

In [7]:
# Converte categorias e verifica níveis, frequências e ausências
colunas_categoricas = ["Species", "Island", "Sex"]
controle_categorias = []
for coluna in colunas_categoricas:
    frequencias_antes = dados[coluna].value_counts(dropna=False).sort_index().to_dict()
    ausentes_antes = int(dados[coluna].isna().sum())
    dados[coluna] = dados[coluna].astype("category")
    frequencias_depois = dados[coluna].value_counts(dropna=False).sort_index().to_dict()
    controle_categorias.append({
        "variavel": coluna,
        "dtype_depois": str(dados[coluna].dtype),
        "ausentes_preservados": ausentes_antes == int(dados[coluna].isna().sum()),
        "frequencias_preservadas": frequencias_antes == frequencias_depois,
    })
pd.DataFrame(controle_categorias)

,variavel,dtype_depois,ausentes_preservados,frequencias_preservadas
0,Species,category,True,True
1,Island,category,True,True
2,Sex,category,True,True


**Tabela 6 - Controle das conversões categóricas.** Espécie, ilha e sexo passam a usar o tipo `category` sem alterar frequências nem ausências. O resultado confirma a preservação dos valores observados, não uma mudança em sua natureza estatística.

## 6. Qualidade dos dados orientada pela pergunta

A qualidade não é uma propriedade absoluta do arquivo. Ela deve ser avaliada em relação à pergunta, às variáveis necessárias e às regras conhecidas da coleta. A análise exploratória permite localizar dados incompletos, inválidos ou inconsistentes, mas verificações internas não provam que uma medição corresponde exatamente ao fenômeno real.

### 6.1 Completude e denominador analítico

**Completude** indica se os valores necessários estão disponíveis. Para a variável $j$, a proporção de ausências pode ser escrita como

$$
p_{	ext{aus},j}=\frac{m_j}{n},
$$

em que $m_j$ é o número de registros ausentes e $n$ é o total de registros considerado. O denominador deve ser declarado, pois pode representar a base completa ou um subconjunto analítico. Um **caso completo** depende do conjunto de variáveis exigido pela pergunta; portanto, não existe uma única contagem de casos completos válida para todas as análises.

Ausência, zero e uma categoria substantiva são situações distintas. Excluir ou preencher valores altera o conjunto analisado e pode modificar sua composição; nesta aula, o diagnóstico não autoriza imputação automática (MCKINNEY, 2017, pp. 209–214).

### 6.2 Validade e consistência

**Validade** verifica se um valor respeita domínio, formato, unidade e intervalo definidos. **Consistência** examina relações entre campos, como datas em ordem possível ou medidas compatíveis com regras documentadas. Um valor dentro do intervalo esperado pode ainda estar incorreto; por isso, o resultado do teste confirma somente a regra avaliada.

### 6.3 Unicidade, unidade de análise e chaves

**Unicidade** exige uma chave coerente com a unidade de análise. Uma linha integralmente repetida é diferente de um identificador que reaparece legitimamente em campanhas distintas. A pergunta correta não é apenas “há valores repetidos?”, mas “quais colunas deveriam identificar unicamente cada observação neste processo de coleta?”.

**Aplicação guiada:** formule uma pergunta envolvendo espécie, massa corporal e sexo. Em seguida, defina as variáveis necessárias, o denominador da completude, as regras de validade e a chave usada para examinar unicidade.

In [8]:
# Calcula quantidade e proporção de ausências por variável
diagnostico_ausencias = (
    pd.DataFrame({
        "ausentes": penguins.isna().sum(),
        "proporcao": penguins.isna().mean(),
    })
    .sort_values(["ausentes", "proporcao"], ascending=False)
)
diagnostico_ausencias

,ausentes,proporcao
Comments,290,0.843023
Delta 15 N (o/oo),14,0.040698
Delta 13 C (o/oo),13,0.037791
Sex,11,0.031977
Culmen Length (mm),2,0.005814
Culmen Depth (mm),2,0.005814
Flipper Length (mm),2,0.005814
Body Mass (g),2,0.005814
studyName,0,0.000000
Sample Number,0,0.000000


**Tabela 7 - Diagnóstico de completude por variável.** `Comments` possui 290 ausências em 344 registros, enquanto sexo apresenta 11 e algumas medidas corporais apresentam 2. A relevância dessas lacunas depende das variáveis exigidas pela pergunta analítica.

In [9]:
# Avalia completude no subconjunto exigido pela pergunta
variaveis_analiticas = ["Species", "Body Mass (g)", "Sex"]
casos_completos = penguins[variaveis_analiticas].notna().all(axis=1)
resumo_completude = pd.DataFrame({
    "medida": ["total de registros", "casos completos", "casos incompletos", "proporção completa"],
    "resultado": [
        len(penguins),
        int(casos_completos.sum()),
        int((~casos_completos).sum()),
        round(casos_completos.mean(), 4),
    ],
})
resumo_completude

,medida,resultado
0,total de registros,344.000
1,casos completos,333.000
2,casos incompletos,11.000
3,proporção completa,0.968


**Tabela 8 - Completude do subconjunto analítico.** Para uma análise que exige simultaneamente espécie, massa corporal e sexo, 333 das 344 observações estão completas. Esse denominador pertence a essa pergunta e não deve ser generalizado para todas as análises do arquivo.

## 7. Pré-processamento, chaves e rastreabilidade

### 7.1 Preservação da base bruta e diagnóstico inicial

A base bruta deve permanecer inalterada. As transformações são aplicadas a uma cópia, permitindo comparar o estado recebido com o estado organizado e refazer o processo quando necessário. Antes de qualquer mudança, registre dimensões, tipos computacionais, domínios, ausências e possíveis duplicidades.

### 7.2 Conversão e padronização controladas

Conversões de datas e categorias devem responder a problemas diagnosticados. A padronização de nomes de colunas pode facilitar o uso do software sem alterar os valores substantivos. Já a padronização de valores exige cuidado adicional: grafias diferentes podem representar a mesma categoria, mas a regra de equivalência precisa ser documentada e conferida.

### 7.3 Duplicatas, chaves simples e chaves compostas

A detecção computacional de duplicatas depende das colunas examinadas (MCKINNEY, 2017, pp. 215–217). Uma **chave simples** usa uma coluna; uma **chave composta** combina duas ou mais. No Palmer Penguins, `Individual ID` isolado pode se repetir entre campanhas. A combinação de `studyName` e `Individual ID` representa uma hipótese de chave mais coerente, que deve ser testada antes de qualquer exclusão.

O método `drop_duplicates()` não decide quais linhas são erros. Ele apenas executa a regra indicada. Usar uma chave incompleta pode eliminar observações válidas e alterar o objeto da análise.

### 7.4 Controles e registro das decisões

Depois de cada transformação, repita controles proporcionais ao risco: número de linhas, ausências, domínios, tipos e unicidade. Registre ao menos cinco elementos:

1. problema diagnosticado;
2. regra aplicada;
3. registros afetados;
4. controle realizado;
5. decisão e justificativa.

Esse registro transforma o pré-processamento em uma sequência auditável. Uma operação disponível no software só deve ser executada quando houver justificativa estatística e uma forma de verificar suas consequências.

In [10]:
# Compara duplicatas integrais e repetições sob chaves alternativas
diagnostico_unicidade = pd.DataFrame([
    {"critério": "linha integral", "registros_em_repetições": int(penguins.duplicated(keep=False).sum())},
    {"critério": "Individual ID", "registros_em_repetições": int(penguins.duplicated(["Individual ID"], keep=False).sum())},
    {"critério": "studyName + Individual ID", "registros_em_repetições": int(penguins.duplicated(["studyName", "Individual ID"], keep=False).sum())},
])
diagnostico_unicidade

,critério,registros_em_repetições
0,linha integral,0
1,Individual ID,268
2,studyName + Individual ID,0


**Tabela 9 - Diagnóstico de unicidade por diferentes critérios.** Não há linhas integralmente duplicadas; 268 registros participam de repetições de `Individual ID`, mas nenhuma repetição permanece na chave composta por campanha e indivíduo. Portanto, usar apenas o identificador individual excluiria observações válidas.

In [11]:
# Padroniza apenas os nomes das colunas na cópia organizada
nomes_antes = dados.columns.to_list()
dados.columns = (
    dados.columns
    .str.strip()
    .str.lower()
    .str.replace(r"[^0-9a-z]+", "_", regex=True)
    .str.strip("_")
)
comparacao_estados = pd.DataFrame({
    "controle": ["linhas na base bruta", "linhas na cópia", "colunas na base bruta", "colunas na cópia"],
    "resultado": [len(penguins), len(dados), penguins.shape[1], dados.shape[1]],
})
comparacao_estados

,controle,resultado
0,linhas na base bruta,344
1,linhas na cópia,344
2,colunas na base bruta,17
3,colunas na cópia,17


**Tabela 10 - Comparação entre a base bruta e a cópia organizada.** A padronização dos nomes preserva as 344 linhas e as 17 colunas. `penguins` mantém os dados recebidos, enquanto `dados` concentra transformações controladas e reprodutíveis.

In [12]:
# Verifica domínios, positividade das medidas e validade das datas
verificacoes = pd.Series({
    "espécies dentro do domínio observado": dados["species"].dropna().isin(penguins["Species"].dropna().unique()).all(),
    "ilhas dentro do domínio observado": dados["island"].dropna().isin(penguins["Island"].dropna().unique()).all(),
    "massas observadas positivas": dados["body_mass_g"].dropna().gt(0).all(),
    "comprimentos de nadadeira positivos": dados["flipper_length_mm"].dropna().gt(0).all(),
    "datas convertidas sem perdas": dados["date_egg"].notna().sum() == penguins["Date Egg"].notna().sum(),
}, name="resultado")
verificacoes.rename_axis("regra").to_frame()

,resultado
regra,
espécies dentro do domínio observado,True
ilhas dentro do domínio observado,True
massas observadas positivas,True
comprimentos de nadadeira positivos,True
datas convertidas sem perdas,True


**Tabela 11 - Verificações básicas de validade e consistência.** As categorias permanecem nos domínios observados, as medidas avaliadas são positivas e a conversão de datas não cria perdas. Cada resultado confirma somente a regra testada e não comprova exatidão da medição ou qualidade absoluta da base.

In [15]:
# Consolida problemas, regras, controles e decisões rastreáveis
registro_decisoes = pd.DataFrame([
    {
        "problema": "data armazenada como texto",
        "regra_aplicada": "conversão para datetime na cópia",
        "registros_afetados": len(dados),
        "controle_realizado": "ausências antes e depois",
        "decisão": "manter conversão; nenhuma perda observada",
    },
    {
        "problema": "categorias armazenadas como object",
        "regra_aplicada": "conversão para category na cópia",
        "registros_afetados": len(dados),
        "controle_realizado": "frequências e ausências",
        "decisão": "manter conversão; valores preservados",
    },
    {
        "problema": "repetição de Individual ID",
        "regra_aplicada": "avaliar chave composta com studyName",
        "registros_afetados": int(penguins.duplicated(["Individual ID"], keep=False).sum()),
        "controle_realizado": "unicidade da chave composta",
        "decisão": "não excluir registros",
    },
    {
        "problema": "nomes de colunas heterogêneos",
        "regra_aplicada": "padronização dos nomes na cópia",
        "registros_afetados": len(nomes_antes),
        "controle_realizado": "dimensões antes e depois",
        "decisão": "manter base bruta separada",
    },
])
registro_decisoes

,problema,regra_aplicada,registros_afetados,controle_realizado,decisão
0,data armazenada como texto,conversão para datetime na cópia,344,ausências antes e depois,manter conversão; nenhuma perda observada
1,categorias armazenadas como object,conversão para category na cópia,344,frequências e ausências,manter conversão; valores preservados
2,repetição de Individual ID,avaliar chave composta com studyName,268,unicidade da chave composta,não excluir registros
3,nomes de colunas heterogêneos,padronização dos nomes na cópia,17,dimensões antes e depois,manter base bruta separada


**Tabela 12 - Registro das transformações e decisões.** Cada intervenção é ligada a um problema diagnosticado, aos registros afetados e a um controle posterior. O registro preserva a rastreabilidade e mostra por que nenhuma imputação ou exclusão foi executada sem evidência.

## 8. Estudo e exercícios

### 8.1 Materiais didáticos

- MORETTIN; BUSSAB (2010), seção 2.1, páginas PDF 26–27: tipos de variáveis.
- BARBETTA; REIS; BORNIA (2010), seção 3.1, páginas PDF 52–53: dados, casos, variáveis, unidades e domínios.
- BRUCE; BRUCE (2019), seção Elementos de Dados Estruturados, páginas PDF 23–26: tipos de dados e representação computacional.
- *Introdução à estatística para ciência de dados*, seção 1.2, páginas PDF 22–27: diagnóstico e pré-processamento.
- MCKINNEY (2017), seção 7.1, páginas PDF 209–214: dados ausentes.
- MCKINNEY (2017), seção 7.2, páginas PDF 215–228: transformações e duplicatas.

### 8.2 Exercícios indicados

No [Banco de questões e provas 2026.2](../apostila/banco_questoes_provas_2026_2.pdf):

- questões 11 e 12, página 10;
- questões 15 e 17, página 11.

Ao responder, justifique a classificação pela natureza da variável e pelo processo de obtenção. Não use apenas o formato visual dos valores ou o tipo indicado pelo software.

## 9. Referências

- BARBETTA, Pedro Alberto; REIS, Marcelo Menezes; BORNIA, Antonio Cezar. *Estatística para cursos de engenharia e informática*. 3. ed. São Paulo: Atlas, 2010.
- BRUCE, Peter; BRUCE, Andrew. *Estatística prática para cientistas de dados: 50 conceitos essenciais*. Rio de Janeiro: Alta Books, 2019.
- *Introdução à estatística para ciência de dados: da exploração dos dados à experimentação contínua com exemplos de código em Python e R*. São Paulo: Casa do Código, 2025.
- MCKINNEY, Wes. *Python for Data Analysis: Data Wrangling with pandas, NumPy, and IPython*. 2. ed. Sebastopol: O'Reilly Media, 2017.
- MORETTIN, Pedro A.; BUSSAB, Wilton O. *Estatística básica*. 6. ed. São Paulo: Saraiva, 2010.
- PINHEIRO, João Ismael D. et al. *Estatística básica: a arte de trabalhar com dados*. Rio de Janeiro: Elsevier, 2009.